pkgs/build-support/go/build-hook.sh

```diff
@@ -0,0 +1,140 @@
+
+goBuildHook() {
+
+  runHook preBuild
+
+  export GOFLAGS
+
+  # currently pie is only enabled by default in pkgsMusl
+  # this will respect the `hardening{Disable,Enable}` flags if set
+  if [[ $NIX_HARDENING_ENABLE =~ "pie" ]]; then
+    prependToVar GOFLAGS "-buildmode=pie"
+  fi
+
+  if [ -z "${allowGoReference-}" ]; then
+    appendToVar GOFLAGS "-trimpath"
+  fi
+
+  if [ -z "${proxyVendor-}" ]; then
+    appendToVar GOFLAGS "-mod=vendor"
+  fi
+
+  # TODO: conditionalize
+  appendToVar ldflags "-buildid="
+
+  exclude='\(/_\|examples\|Godeps\|testdata'
+  if [[ -n "$excludedPackages" ]]; then
+    IFS=' ' read -r -a excludedArr <<<$excludedPackages
+    printf -v excludedAlternates '%s\\|' "${excludedArr[@]}"
+    excludedAlternates=${excludedAlternates%\\|} # drop final \| added by printf
+    exclude+='\|'"$excludedAlternates"
+  fi
+  exclude+='\)'
+
+  buildGoDir() {
```

Comment:
IIRC, Bash doesn't support nested function declaration. Let's flatten the definition of these Bash functions.